ggml-cuda.cu

```diff
@@ -5987,6 +6035,527 @@ static  __global__ void im2col_f32_f16(
+            ss[j*T + i] = 0.0;
+        }
+    }
+
+    __syncthreads();
+
+    {
+        float S[Q];
+        float M[Q];
+
+        for(int i = 0; i < Q;i ++) {
+            S[i] = 0.0f;
+            M[i] = -INFINITY;
+        }
+
+        // assume K and V are same shape
+        const int ne22 = ne12;
+        const int ne23 = ne13;
+
+        const int nb21 = nb11;
+        const int nb22 = nb12;
+        const int nb23 = nb13;
+
+        // broadcast
+        const int rk2 = ne02/ne12;
+        const int rk3 = ne03/ne13;
+
+        const int rv2 = ne02/ne22;
+        const int rv3 = ne03/ne23;
+
+        // k indices
+        const int ik2 = iq2 / rk2;
+        const int ik3 = iq3 / rk3;
+
+        // v indices
+        const int iv2 = iq2 / rv2;
+        const int iv3 = iq3 / rv3;
+
+        // load the queries from shared memory into local memory
+        half16x16_a mq[Q16][D16];
+        for (int64_t j = 0; j < Q16; ++j) {
+            for (int64_t i = 0; i < D16; ++i) {
+                nvcuda::wmma::load_matrix_sync(mq[j][i], sq + 16*j*T + i*16, T);
+            }
+        }
+
+        const int64_t ir = iq3*ne02*ne01 + iq2*ne01 + iq1;
+
+        // pointer to the mask
+        const float * mp = (const float *) (mask + (ir%ne31)*nb31);
+
+        // loop over the KV cache
+        // each simdgroup handles blocks of Q rows and C columns
+        for (int64_t ic = C*warp_id; ic < ne11; ic += C*num_warps) {
+            // Q*K^T
+            {
+                for (int cc = 0; cc < C/16; ++cc) {
+                    half16x16_acc mqk[Q16];
+                    for (int64_t j = 0; j < Q16; ++j) {
+                        nvcuda::wmma::fill_fragment(mqk[j], 0);
+                    }
+
+                    const half * pk = (const half *) ((const char *) k + ((ic + 16*cc)*nb11 + ik2*nb12 + ik3*nb13));
+
+                    for (int64_t i = 0; i < D16; ++i) {
+                        half16x16_bT mk; // transposed key
+                        nvcuda::wmma::load_matrix_sync(mk, pk + i*16, nb11/sizeof(half)); // transpose
+
+                        for (int64_t j = 0; j < Q16; ++j) {
+                            nvcuda::wmma::mma_sync(mqk[j], mq[j][i], mk, mqk[j]);
+                        }
+                    }
+
+                    // mqk = mqk*scale + mask
+                    for (int64_t j = 0; j < Q16; ++j) {
+                        const float* msk_p = mp + 16*j*(nb31/sizeof(float)) + ic + 16*cc;
+                        int64_t msk_ne_row = nb31/sizeof(float);
+                        for (uint32_t i = 0; i < mqk[j].num_elements; i++) {
+                            int msk_col = i % 16;
+                            int msk_row = i / 16;
+                            mqk[j].x[i] = __float2half(scale * __half2float(mqk[j].x[i]) + msk_p[msk_col + msk_row*msk_ne_row]);
+                        }
+                        nvcuda::wmma::store_matrix_sync(ss + 16*j*T + 16*cc, mqk[j], T, nvcuda::wmma::mem_col_major);
```

Comment:
I could be misunderstanding how these functions work, but I think all `store_matrix_sync` and `load_matrix_sync` should use `mem_row_major` and only the `mk` matrix should load using `mem_col_major`

Reply:
For unfused general matrix multiplication kernels, to get correct results for `src0` as fragment a and `src1` as fragment b, you have to load `src0`/fragment a as row major, `src1`/fragment b as col major, and you have to store the result as col major. But I did **not** analyze the exact memory layout needed for a fused kernel that does multiple matrix multiplications at once.